src/openrct2/core/FileScanner.cpp

```diff
@@ -23,335 +23,156 @@
-        if (count > 0)
+        if (MatchWildcard(fileName.c_str(), pattern.c_str()))
         {
-            for (int32_t i = 0; i < count; i++)
-            {
-                const struct dirent* node = namelist[i];
-                if (!String::Equals(node->d_name, ".") && !String::Equals(node->d_name, ".."))
-                {
-                    DirectoryChild child = CreateChild(path.c_str(), node);
-                    children.push_back(child);
-                }
-                free(namelist[i]);
-            }
-            free(namelist);
+            return true;
         }
     }
+    return false;
+}
 
-private:
-    static int32_t FilterFunc(const struct dirent* d)
-    {
-        return 1;
-    }
+std::vector<std::string> FileScannerBase::GetPatterns(const std::string& delimitedPatterns)
+{
+    std::vector<std::string> patterns;
 
-    static DirectoryChild CreateChild(const utf8* directory, const struct dirent* node)
+    const utf8* start = delimitedPatterns.c_str();
+    const utf8* ch = start;
+    utf8 c;
+    do
     {
-        DirectoryChild result;
-        result.Name = std::string(node->d_name);
-        if (node->d_type == DT_DIR)
+        c = *ch;
+        if (c == '\0' || c == ';')
         {
-            result.Type = DIRECTORY_CHILD_TYPE::DC_DIRECTORY;
-        }
-        else
-        {
-            result.Type = DIRECTORY_CHILD_TYPE::DC_FILE;
-
-            // Get the full path of the file
-            size_t pathSize = String::SizeOf(directory) + 1 + String::SizeOf(node->d_name) + 1;
-            utf8* path = Memory::Allocate<utf8>(pathSize);
-            String::Set(path, pathSize, directory);
-            Path::Append(path, pathSize, node->d_name);
-
-            struct stat statInfo
-            {
-            };
-            int32_t statRes = stat(path, &statInfo);
-            if (statRes != -1)
+            size_t length = static_cast<size_t>(ch - start);
+            if (length > 0)
             {
-                result.Size = statInfo.st_size;
-                result.LastModified = statInfo.st_mtime;
-
-                if (S_ISDIR(statInfo.st_mode))
-                {
-                    result.Type = DIRECTORY_CHILD_TYPE::DC_DIRECTORY;
-                }
+                std::string newPattern = std::string(start, length);
+                patterns.push_back(newPattern);
             }
-
-            Memory::Free(path);
+            start = ch + 1;
         }
-        return result;
-    }
-};
+        ch++;
+    } while (c != '\0');
 
-#endif // defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
+    patterns.shrink_to_fit();
+    return patterns;
+}
 
 IFileScanner* Path::ScanDirectory(const std::string& pattern, bool recurse)
 {
-#ifdef _WIN32
-    return new FileScannerWindows(pattern, recurse);
-#elif defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
-    return new FileScannerUnix(pattern, recurse);
-#endif
+    return Platform::ScanDirectory(pattern, recurse);
```

Comment:
Add a Todo saying replace with std::unique_ptr and create an issue to track it.

Reply:
Done